lib/s3-routes.js

```diff
@@ -0,0 +1,2719 @@
+                    var s3MetaKey = 'x-amz-meta-' + lowerKey.substring(2);
+                    req.log.debug({
+                        originalKey: key,
+                        convertedKey: s3MetaKey,
+                        value: value
+                    }, 'S3_DEBUG: Converting HEAD '+
+                       'metadata header to S3 format');
+                    return (originalSet.call(this, s3MetaKey, value));
+                }
+            }
+            return (originalSet.call(this, key, value));
+        };
+
+        // Call the original Manta head object handler
+        var mantaHandlerChain = buckets.headBucketObjectHandler();
+        executeMiddlewareChain(mantaHandlerChain, req, res, function (error) {
+            // Always terminate the route chain for S3 requests to
+            // prevent double execution
+            if (error) {
+                next(error);
+            } else {
+                next(false); // Stop route processing
+            }
+        });
+    };
+}
+
+/**
+ * S3 Delete Object Handler
+ */
+function s3DeleteBucketObjectHandler() {
+    return function s3DeleteBucketObject(req, res, next) {
+        // With regex routes, captured groups are in req.params array
+        var bucketName = req.params[0] || req.params.bucket;
+        var objectPath = req.params[1] || req.params['*'] || '';
+
+        // Check if this is a tagging operation
+        if (req.query && req.query.tagging !== undefined) {
+            req.log.debug({
+                bucket: bucketName,
+                object: objectPath,
+                query: req.query
+            }, 'S3_DEBUG: DELETE Object tagging operation detected');
+            s3DeleteObjectTaggingHandler()(req, res, next);
+            return;
+        }
+
+        req.log.debug({
+            bucket: bucketName,
+            object: objectPath
+        }, 's3DeleteBucketObject: processing S3 delete object request');
+
+        if (!req.caller || !req.caller.account) {
+            var err = new Error('Authentication required');
+            err.statusCode = 401;
+            next(err);
+            return;
+        }
+
+        // Convert S3 parameters to Manta format
+        req.params.account = req.caller.account.login;
+        req.params.bucket_name = bucketName;
+        req.params.object_name = objectPath;
+
+        // Call the original Manta delete object handler
+        var mantaHandlerChain = buckets.deleteBucketObjectHandler();
+        executeMiddlewareChain(mantaHandlerChain, req, res, function (error) {
+            // Always terminate the route chain for S3 requests to
+            // prevent double execution
+            if (error) {
+                next(error);
+            } else {
+                next(false); // Stop route processing
+            }
+        });
+    };
+}
+
+/**
+ * S3 bulk delete handler for POST /bucket?delete requests
+ */
+
+/* BEGIN JSSTYLED */
+function s3DeleteBucketObjectsHandler() {
```

Comment:
This is a big and deep function.  Can you factor things out to make it more readable?